src/stores/index.js

```diff
@@ -851,23 +896,38 @@ const Store = Reflux.createStore({
     // Set the connection's app name to the electron app name of Compass.
     connection.appname = electron.remote.app.getName();
 
-    const dataService = new DataService(connection);
+    if (!this.state.isConnecting) {
+      // The connection attempt might have been cancelled while
+      // we were parsing the connection information, so we return here.
+      return;
+    }
 
     try {
-      const runConnect = promisify(dataService.connect.bind(dataService));
-      const connectedDataService = await runConnect();
+      const dataService = new DataService(connection);
+      this.connectingConnectionAttempt = createConnectionAttempt(dataService);
+
+      const connectedDataService = await this.connectingConnectionAttempt.connect();
+
+      this.connectingConnectionAttempt = null;
+
+      if (!connectedDataService || !this.state.isConnecting) {
+        return;
+      }
 
       const currentConnection = this.state.currentConnection;
```

Comment:
why do we re-read the connection from the state instead of using `connection` from above here, that shouldn't that be the source of truth?

Reply:
Chatted about this on video - it's because the connection model passed to this method isn't always the same as the one on the state - the connect with connection string for instance mutates a new connection derived from the state which isn't the same. Something we'll address in future and clean up since it's probably causing bugs or could cause bugs.